jgrapht-core/src/main/java/org/jgrapht/alg/tour/HamiltonianCycleAlgorithmBase.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * (C) Copyright 2019-2019, by Peter Harman and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * See the CONTRIBUTORS.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the
+ * GNU Lesser General Public License v2.1 or later
+ * which is available at
+ * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht.alg.tour;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.jgrapht.Graph;
+import org.jgrapht.GraphPath;
+import org.jgrapht.alg.interfaces.HamiltonianCycleAlgorithm;
+import org.jgrapht.graph.GraphWalk;
+
+/**
+ * Base class for TSP solver algorithms.
+ *
+ * <p>
+ * This class provides implementations of utilities for TSP solver classes.
+ * </p>
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Peter Harman
+ */
+public abstract class HamiltonianCycleAlgorithmBase<V, E> implements HamiltonianCycleAlgorithm<V, E> {
+
+    /**
+     * Transform from a List representation to a graph path.
```

Comment:
I agree that would be more flexible, but GraphWalk requires a List<V>, so we save creation of a new List by reusing the one passed in? Of course if LinkedList implemented List (#830 ) it would also fix it?

Reply:
I meant `java.util.LinkedList` not our own.

Reply:
OK, I thought you meant the function signature but you meant the for-loop, I've made this change